### PR TITLE
Localtuya 4.0

### PR DESCRIPTION
**Pull request by rospogrigio**
Major change in Localtuya. Here's what's new:
- one single integration instead of one integration for each device
- introduced Cloud API connection (for local key retrieval)
- auto-update of local keys (in case of re-pairing of a device)
- introduced diagnostics, both for the Integration and for each device
- **removed YAML configuration** (since it's being deprecated almost everywhere)

### CODE REVIEW

File: custom_components/localtuya/cloud_api.py
Comment (CloCkWeRX) [inline custom_components/localtuya/cloud_api.py:1]: Oh, nice.

**Comment by hiagocosta** [comment]
when 4.0 will be released to HACS?

**Comment by rospogrigio** [comment]
> when 4.0 will be released to HACS?

Quite shortly I believe, I am just fixing the last issues I'm finding, I fixed one related to the diagnostics and I noticed something not working correctly when editing a device, then I'll be ready for release I guess. You can still start using the PR in the meantime, I believe it's pretty stable apart from the "Edit device" feature.

**Comment by rospogrigio** [comment]
@CloCkWeRX I believe this PR is ready for prime time now, can you please test it and review it? I'd like to have an external review before I merge everything an launch the new release.
Thank you!

**Comment by EnsignR** [comment]
I've updated from the commits 6 days ago, [ef50285](https://github.com/rospogrigio/localtuya/commit/ef50285c33bcfbb1f8a8e0942c8f5caf751e85a7) to the current and all the devices I set up with the earlier update work, but if I try to add a new device manually...

`Unknown error occurred`

```
...
  File "/config/custom_components/localtuya/config_flow.py", line 543, in async_step_configure_device
    data_schema=schema,
UnboundLocalError: local variable 'schema' referenced before assignment
```

**Comment by rospogrigio** [comment]
> I've updated from the commits 6 days ago, [ef50285](https://github.com/rospogrigio/localtuya/commit/ef50285c33bcfbb1f8a8e0942c8f5caf751e85a7) to the current and all the devices I set up with the earlier update work, but if I try to add a new device manually...
> 
> `Unknown error occurred`
> 
> ```
> ...
>   File "/config/custom_components/localtuya/config_flow.py", line 543, in async_step_configure_device
>     data_schema=schema,
> UnboundLocalError: local variable 'schema' referenced before assignment
> ```

Ooops you are right, I have never tested that part... ok let me check it...

**Comment by rospogrigio** [comment]
OK I seem to have fixed it @EnsignR , could you please check it?

**Comment by EnsignR** [comment]
Can confirm the dialog now opens _here_ too. Cheers. 👍

File: custom_components/localtuya/common.py
Comment (CloCkWeRX) [inline custom_components/localtuya/common.py:250]: Is this at the right level? IE, can self._interface and self._disconnect_task be None; yet this still logs that it disconnected; even if it did nothing?

File: info.md
Comment (CloCkWeRX) [inline info.md:9]:
```suggestion
This custom integration updates device status via pushing updates instead of polling, so status updates are fast (even when manually operated).
```

File: info.md
Comment (CloCkWeRX) [inline info.md:49]:
```suggestion
**NOTE: starting from v4.0.0, configuration using YAML files is unsupported. The integration can only be configured using the config flow.**
```

File: info.md
Comment (CloCkWeRX) [inline info.md:133]:
```suggestion
2) Access the voltage/current/current_consumption attributes of a switch, and define template sensors
```

Unsure if this is a markdown rendering behaviour, or this should be 1); 2); 3)

**Review by CloCkWeRX — APPROVED** [review]
Reviewed what I got of the higher level change, and then focused more on style/syntax/basics around logic.

I think broadly the changes make sense (config flow, etc); though I don't have the greatest mental model of how this all hangs together.

I have not yet tested locally on my setup.

Most feedback is minor documentation or possible dead code related.

File: custom_components/localtuya/__init__.py
Comment (CloCkWeRX) [inline custom_components/localtuya/__init__.py:306]: May want to consider the changes in https://github.com/rospogrigio/localtuya/pull/842 being applied here to avoid
```
Detected integration that uses deprecated `async_get_registry` to access entity registry, use async_get instead
```

File: custom_components/localtuya/__init__.py
Comment (CloCkWeRX) [inline custom_components/localtuya/__init__.py:340]: Also here, re async_get_registry

File: custom_components/localtuya/common.py
Comment (CloCkWeRX) [inline custom_components/localtuya/common.py:121]: After upgrade; in my logs:
```
Logger: homeassistant
Source: custom_components/localtuya/common.py:121
Integration: LocalTuya ([documentation](https://github.com/rospogrigio/localtuya/), [issues](https://github.com/rospogrigio/localtuya/issues))
First occurred: 9:06:13 PM (7 occurrences)
Last logged: 9:06:15 PM

Error doing job: Exception in callback _SelectorDatagramTransport._read_ready()
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.9/asyncio/selector_events.py", line 1029, in _read_ready
    self._protocol.datagram_received(data, addr)
  File "/config/custom_components/localtuya/discovery.py", line 70, in datagram_received
    self.device_found(decoded)
  File "/config/custom_components/localtuya/discovery.py", line 79, in device_found
    self._callback(device)
  File "/config/custom_components/localtuya/__init__.py", line 103, in _device_discovered
    entry = async_config_entry_by_device_id(hass, device_id)
  File "/config/custom_components/localtuya/common.py", line 121, in async_config_entry_by_device_id
    if device_id in entry.data[CONF_DEVICES]:
KeyError: 'devices'
```

Unclear exactly what's happening / may be a mistake on my part.

File: custom_components/localtuya/__init__.py
Comment (CloCkWeRX) [inline custom_components/localtuya/__init__.py:234]: Probably as I haven't *done* the cloud configuration to set up all of this; I get errors after upgrade.

May be worth raising errors if not .is_configured from a UX point of view
```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 335, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/localtuya/__init__.py", line 234, in async_setup_entry
    region = entry.data[CONF_REGION]
KeyError: 'region'
```

File: custom_components/localtuya/__init__.py
Comment (CloCkWeRX) [inline custom_components/localtuya/__init__.py:241]: Ditto re not yet configured warnings - this is possibly a bit misleading.

```
Logger: custom_components.localtuya
Source: custom_components/localtuya/__init__.py:241
Integration: LocalTuya ([documentation](https://github.com/rospogrigio/localtuya/), [issues](https://github.com/rospogrigio/localtuya/issues))
First occurred: 9:06:18 PM (1 occurrences)
Last logged: 9:06:18 PM

Cloud API connection failed: Error 1114: your ip(149.167.167.219) don't have access to this API
```

File: custom_components/localtuya/strings.json
Comment (CloCkWeRX) [inline custom_components/localtuya/strings.json:48]: Can't find the exact right spot; but on this UI - help link just points to the github repo.
image

Might be better to point to the README or a specific relevant subsection (not sure how thought)

File: custom_components/localtuya/common.py
Comment (rospogrigio) [reply custom_components/localtuya/common.py:250]: Well, I just want to make sure that that method is called and ends, regardless of whether the device was actually connected at that time, so I'd keep it like this.

File: custom_components/localtuya/__init__.py
Comment (rospogrigio) [reply custom_components/localtuya/__init__.py:306]: Gosh, I didn't know about this. OK I'll use the new call.

File: custom_components/localtuya/common.py
Comment (rospogrigio) [reply custom_components/localtuya/common.py:121]: This is strange because your config entry should have been migrated and after the migration that key should exist. Maybe there's a race condition with the discovery? I'll check this.

File: custom_components/localtuya/__init__.py
Comment (rospogrigio) [reply custom_components/localtuya/__init__.py:241]: I know, but that message is coming from the Cloud, I am just forwarding it. This is what it usually replies if you input the wrong credentials, the reply is weird but I would keep it like this.

File: README.md
Comment (rospogrigio) [reply README.md:65]: Good point.

File: custom_components/localtuya/__init__.py
Comment (CloCkWeRX) [inline custom_components/localtuya/__init__.py:140]: Also for some reason; *after* I had done migration and successful cloud setup; this appears. Possibly not initalised to hass.data[DOMAIN][TUYA_DEVICES] = {} or similar at some point by an async method?
Something else?
Unclear.

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.9/asyncio/selector_events.py", line 1029, in _read_ready
    self._protocol.datagram_received(data, addr)
  File "/config/custom_components/localtuya/discovery.py", line 70, in datagram_received
    self.device_found(decoded)
  File "/config/custom_components/localtuya/discovery.py", line 79, in device_found
    self._callback(device)
  File "/config/custom_components/localtuya/__init__.py", line 140, in _device_discovered
    device = hass.data[DOMAIN][TUYA_DEVICES][device_id]
KeyError: 'tuya_devices'
```

Comment (rospogrigio) [reply custom_components/localtuya/__init__.py:140]: OK I believe we do have a race condition here... does it persist after you restart HA?

File: custom_components/localtuya/strings.json
Comment (EnsignR) [reply custom_components/localtuya/strings.json:21]: They're displaying for me if I reconfigure the account originally set up with [ef50285](https://github.com/rospogrigio/localtuya/commit/ef50285c33bcfbb1f8a8e0942c8f5caf751e85a7)
image

Comment (rospogrigio) [reply custom_components/localtuya/strings.json:21]: This is fixed now, check it out...

File: custom_components/localtuya/strings.json
Comment (rospogrigio) [reply custom_components/localtuya/strings.json:48]: OK I think I found it, it's the "documentation" field in the manifest.json . It's not something you can make window-specific (I mean, you only have one link, you can't have a different link to different subsection for every configuration step, which is what you are proposing, probably). So I'd keep it like this: it points to the repo, which is what is commonly done in all custom components.

File: custom_components/localtuya/common.py
Comment (racu10) [inline custom_components/localtuya/common.py:75]: why this? maybe directly` if entities_to_setup`

Comment (rospogrigio) [reply custom_components/localtuya/common.py:75]: Well, this is because I'm not a native python programmer (I am more a C++ coder), so I don't know all these tricks... thank you for the suggestion! 😉

File: custom_components/localtuya/cloud_api.py
Comment (rospogrigio) [reply custom_components/localtuya/cloud_api.py]: Yeah, it's just some output kept disabled for the moment... but ok, I've replaced these with log.debug which is better.

**Comment by CloCkWeRX** [comment]
Updated round of testing (WIP/will keep editing comments here)...

Merged to https://github.com/CloCkWeRX/localtuya-experimental
Cleared logs
Restarted....

File: custom_components/localtuya/__init__.py
Comment (CloCkWeRX) [reply custom_components/localtuya/__init__.py:141]: Solved

**Comment by CloCkWeRX** [comment]
Alright, seems good to go IMO on both this month and last month's HA install. @rospogrigio merge away!

**Comment by rospogrigio** [comment]
> Alright, seems good to go IMO on both this month and last month's HA install. @rospogrigio merge away!

Not yet, I realized that you can't actually add the Integration without configuring the cloud credentials, I am adding an option for this, then I think it's merging time. Thank you for the precious review!

**Comment by rospogrigio** [comment]
OK @CloCkWeRX I think I'm ready for release, can you please do some more checks and approve the review? I just added a button to choose not to setup the Cloud API account. Thank you!

File: README.md
Comment (rospogrigio) [reply README.md]: Fixed introducing a checkbox that allows to choose not to configure the Cloud account.

**Comment by rospogrigio** [comment]
@CloCkWeRX can you please review the new commits and approve it when you find the time? I will then release the new code.
Thank you!

**Comment by CloCkWeRX** [comment]
LGTM - I can't re-add an approving review/I'm not core so it doesn't pass the "At least 1 approving review is required by reviewers with write access." check, but merge away

**Review by smarthomefamilyverrips — APPROVED** [review]
LGTM

**Comment by regevbr** [comment]
Nice! When can we expect a release?

**Comment by rospogrigio** [comment]
> Nice! When can we expect a release?

Just did it, check it and enjoy! 😉

**Comment by regevbr** [comment]
Works flawlessly for me :-) thank you so much!!!

**Pull request merged**
